### ISSUE TITLE
Can't tell failure reason from the log

### ISSUE THREAD

**Issue by joker-eph** [google/llvm-premerge-checks#176]
This [diff](https://reviews.llvm.org/harbormaster/build/64785/1/) has been built twice and failed both times, but from the log it isn't clear what failed:

https://results.llvm-merge-guard.org/amd64_debian_testing_clang8-9444/console-log.txt
https://results.llvm-merge-guard.org/amd64_debian_testing_clang8-9447/console-log.txt

**Comment by rengolin** [comment]
There was only one test, which passed:

    Testing Time: 489.87s
      Unsupported Tests:   674
      Expected Passes  : 65160
      Expected Failures:   242

And the clang-format also had no issues:

    clang-format did not modify any files

Using `set -e` can "fail" if one of the conditions in an if fails even if it was supposed to. It gets really hard to know what failed just by looking at the logs. You should really set a trap or check for $? or something more robust and meaningful.

**Comment by ChristianKuehnel** [comment]
looking at the first log file, `ninja all` failed:
```
[4943/5766] Building CXX object tools/mlir/tools/mlir-opt/CMakeFiles/obj.MLIRMlirOptMain.dir/mlir-opt.cpp.o
FAILED: tools/mlir/tools/mlir-opt/CMakeFiles/obj.MLIRMlirOptMain.dir/mlir-opt.cpp.o 
CCACHE_DIR=/mnt/disks/ssd0/ccache CCACHE_MAXSIZE=20G CCACHE_CPP2=yes CCACHE_HASHDIR=yes /usr/bin/ccache /usr/bin/clang++  -DGTEST_HAS_RTTI=0 -DMLIR_CUDA_CONVERSIONS_ENABLED=1 -D_DEBUG -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Itools/mlir/tools/mlir-opt -I/mnt/disks/ssd0/agent/workspace/amd64_debian_testing_clang8/mlir/tools/mlir-opt -I/usr/include/libxml2 -Iinclude -I/mnt/disks/ssd0/agent/workspace/amd64_debian_testing_clang8/llvm/include -I/mnt/disks/ssd0/agent/workspace/amd64_debian_testing_clang8/mlir/include -Itools/mlir/include -gmlt -fPIC -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wstring-conversion -fdiagnostics-color -ffunction-sections -fdata-sections -O3     -fno-exceptions -fno-rtti -UNDEBUG -std=c++14 -MD -MT tools/mlir/tools/mlir-opt/CMakeFiles/obj.MLIRMlirOptMain.dir/mlir-opt.cpp.o -MF tools/mlir/tools/mlir-opt/CMakeFiles/obj.MLIRMlirOptMain.dir/mlir-opt.cpp.o.d -o tools/mlir/tools/mlir-opt/CMakeFiles/obj.MLIRMlirOptMain.dir/mlir-opt.cpp.o -c /mnt/disks/ssd0/agent/workspace/amd64_debian_testing_clang8/mlir/tools/mlir-opt/mlir-opt.cpp
In file included from /mnt/disks/ssd0/agent/workspace/amd64_debian_testing_clang8/mlir/tools/mlir-opt/mlir-opt.cpp:16:
In file included from /mnt/disks/ssd0/agent/workspace/amd64_debian_testing_clang8/mlir/include/mlir/InitAllDialects.h:24:
/mnt/disks/ssd0/agent/workspace/amd64_debian_testing_clang8/mlir/include/mlir/Dialect/Linalg/IR/LinalgOps.h:84:10: fatal error: 'mlir/Dialect/Linalg/IR/LinalgStructuredOpsInterfaces.h.inc' file not found
#include "mlir/Dialect/Linalg/IR/LinalgStructuredOpsInterfaces.h.inc"
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```

**Comment by rengolin** [comment]
Weird. If the build fails, how come it can run and pass the tests?

**Comment by rengolin** [comment]
Also, this is a wider issue. We can't see any of that from the Phab page. Having a pre-merge test that fails without any further information is worse than having a post-merge buildbot fail with all the information available.

If we're going to have pre-merge checks, they need to either be publicly accessible or propagate the error messages concisely and accurately, as well as not be flaky.

I can't imagine how Matt's change has triggered the error above.

**Comment by joker-eph** [comment]
> I can't imagine how Matt's change has triggered the error above.

The build was already broken.

Actually this issue is interesting: the script detected that the build failed (running `ninja`), but still continues and try to run the tests (`ninja check-all`). Because the error is non deterministic (missing dependency triggering an issue only in parallel builds), then running `ninja check-all` passed afterwards.

**Comment by rengolin** [comment]
Interesting! I see, but that means what "passed" was an earlier build, no?

**Comment by joker-eph** [comment]
Not sure what you mean.

I think the script invoked separately the build and the test. So the build “failed” but when running the tests, ninja just resume the build where it failed (but this time it’ll pass)

**Comment by rengolin** [comment]
Of course! Ok, so the test is an actual pass.

When dealing with flaky tests like that, some frameworks run the entire test again N times until it passes or it gives up, but only showing the last result. Otherwise, we get weird cases like this.

Also, how can I reach the logs from the Phab page?

**Comment by ChristianKuehnel** [comment]
There are several things to explain here:

1. We're looking into replacing the (hidden) Jenkins builds with Buildkite. But that's work in progress. This would give you a better visibility of the build status.

1. We're running the linters and builds in the same [pipeline](https://github.com/google/llvm-premerge-checks/blob/master/Jenkins/BETA-phabricator-linux/Jenkinsfile) at the moment. This means that we continue building even if one of the steps fails so that you still get lint results even if compilation fails. We should probably split that up when moving to a new infrastructure #177.

1. Phabricator shows the build correctly as "failed". So there was no risk of integrating a failing build. Phabricator does not offer a whole lot of options for giving detailed build feedback, this is why we're generating the [summary](https://results.llvm-merge-guard.org/amd64_debian_testing_clang8-9444/summary.html). And that also correctly classified the build "ninja all failed". This summary is meant as the first place for users to look for the build status. 

1. Pre-merge tests are optional at the moment for several reasons, reliability of the entire infrastructure being one of them. The pre-merge tests are not a paid-for service. They are right now a best-effort side project for a few community members, as with most LLVM infrastructure. So we'd be happy to get people onboard to improve them.

1. Overall pre-merge testing would be 10x simpler if LLVM would be using Github Pull Requests instead of Phabricator. For PR you get tons of nice tools for free. But thats an entirely different topic...

**Comment by rengolin** [comment]
Hi Christian, thanks for the detailed feedback.
1. Awesome, thanks!
2. Splitting is probably a good idea.
3. It would be the correct classification if the build had actually failed. I know it's not possible to distinguish in Phab, but to be able to rely on it, the build must be stable. In LLVM, unstable buildbots get turned off for until they can be proven stable again.
4. Been there, done that. :) I'm not blaming the infrastructure, just trying to help make it better. I think a simple internal loop on each stage (3 times is enough) until it's green would go a long way of getting rid of the instability.
5. Don't get me started! :D

**Comment by joker-eph** [comment]
> Also, how can I reach the logs from the Phab page?

![Screen Shot 2020-05-06 at 11 50 38 AM](https://user-images.githubusercontent.com/3372300/81216489-d54e5f80-8f8f-11ea-988e-d1da5a4a1130.png)

You have a link to the build log here.


> We're running the linters and builds in the same pipeline at the moment. This means that we continue building even if one of the steps fails so that you still get lint results even if compilation fails. 

What's missing is that the bottom of the log a recap of all the steps and an indication of which one succeeded/failed.
The same script implemented on GitHub PR would have the exact same problem in this case.

**Comment by rengolin** [comment]
Thanks @joker-eph, I didn't realise those things were links.

**Comment by ChristianKuehnel** [comment]
@joker-eph improving the provided log files is a good idea, I just added #178 for that. Iw would also be nice to have some better way of serving the logs than plain text, I haven't found anything in the open source world so far.

On Github PR, I would just refer to the Buildkite UI, that contains all relevant data (can't add link as buildkite is down so much for stability there 😉 ).

**Comment by ChristianKuehnel** [comment]
@rengolin 

3. I haven't seen much flakiness in builds. We had some bugs in the infrastructure we're trying to shake out over time, but actual random failures are rare in my observation.

4. Re-running the entire test build is way to expensive with CMake in CPU resources and wall time. We already have much longer build times than I would like to have, even with ccache/sccache in place. If you want such features, we need a build system that handles incremental builds and changes in dependencies better.

**Comment by rengolin** [comment]
I don't mean the entire build, just the part that failed. In fact, this is what just happened, when `ninja check` pulled the rest of the build, but without the semantics that it made the build actually pass.

Of course, a much better fix is to get rid of those issues in the first place. If `.inc` files are delayed before they get included, it means the files that include them aren't dependencies of the table-gen process that created them.
